xds/internal/server/rds_handler_test.go

```diff
@@ -20,325 +20,391 @@ package server
+			route2: defaultRouteConfigUpdate,
+		},
 	}
+	verifyUpdateFromChannel(ctx, t, updateCh, wantUpdate)
 
-	// Invoke the callback with an update for route 2. This shouldn't cause the
-	// handler to write an update, as it has not received RouteConfigurations
-	// for every RouteName.
-	rdsUpdate2 := xdsresource.RouteConfigUpdate{}
-	fakeClient.InvokeWatchRouteConfigCallback(route2, rdsUpdate2, nil)
+	// Update the handler with a route that was already there and a new route.
+	rh.updateRouteNamesToWatch(map[string]bool{route2: true, route3: true})
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{route2, route3})
 
-	// The RDS Handler should not send an update.
+	// The handler should not send an update.
 	sCtx, sCtxCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
 	defer sCtxCancel()
 	select {
-	case <-ch:
+	case <-updateCh:
 		t.Fatalf("RDS Handler wrote an update to updateChannel when it shouldn't have, as each route name has not received an update yet")
 	case <-sCtx.Done():
 	}
 
-	// Invoke the callback with an update for route 3. This should cause the
-	// handler to write an update, as it has received RouteConfigurations for
-	// every RouteName.
-	rdsUpdate3 := xdsresource.RouteConfigUpdate{}
-	fakeClient.InvokeWatchRouteConfigCallback(route3, rdsUpdate3, nil)
-	// The RDS Handler should then update the listener wrapper with an update
-	// with two route configurations, as both route names the RDS Handler handles
-	// have received an update.
-	rhuWant := map[string]xdsresource.RouteConfigUpdate{route2: rdsUpdate2, route3: rdsUpdate3}
-	select {
-	case rhu := <-rh.updateChannel:
-		if diff := cmp.Diff(rhu.updates, rhuWant); diff != "" {
-			t.Fatalf("got unexpected route update, diff (-got, +want): %v", diff)
-		}
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for the rds handler update to be written to the update buffer.")
+	// Configure the management server with the third resource.
+	routeResource3 := routeConfigResourceForName(route3)
+	resources.Routes = append(resources.Routes, routeResource3)
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
 	}
-	// Close the rds handler. This is meant to be called when the lis wrapper is
-	// closed, and the call should cancel all the watches present (for this
-	// test, two watches on route2 and route3).
-	rh.close()
-	if err = waitForFuncWithNames(ctx, fakeClient.WaitForCancelRouteConfigWatch, route2, route3); err != nil {
-		t.Fatalf("Error while waiting for names: %v", err)
+
+	// Verify that the update is pushed to the handler's update channel.
+	wantUpdate = rdsHandlerUpdate{
+		updates: map[string]xdsresource.RouteConfigUpdate{
+			route2: defaultRouteConfigUpdate,
+			route3: defaultRouteConfigUpdate,
+		},
 	}
+	verifyUpdateFromChannel(ctx, t, updateCh, wantUpdate)
+
+	// Close the rds handler and verify that the watch is canceled.
+	rh.close()
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{})
```

Comment:
This is done at the end of every test. I guess I'm fine with it because it closes from different state spaces. Also isn't it more than one watch that gets cancelled haha?

Reply:
> Also isn't it more than one watch that gets cancelled haha?

In some tests, it is more than one watch and in some it it one. I think the comment conveys the message enough.